### PR TITLE
add pair/tuple conversions

### PR DESCRIPTION
**Pull request by theodelrieu**
I left a question in the code: Should we impose that the JSON array size must be equal to the number of tuple/pair elements?

If not, some elements in the JSON could be discarded:

```cpp
json j = std::make_tuple(1, 2, 3, 4);
auto p = j.get<std::pair<int, int>>();
```

Some users might want to discard values on purpose, so I would say we should not check the size at all.

### CODE REVIEW

**Comment by nlohmann** [comment]
What would be the behavior if I try to convert `[1]` to a `std::pair<int,int>`?

**Comment by theodelrieu** [comment]
It would throw, I use `at` in the implementation

**Comment by nlohmann** [comment]
Right, and it makes sense to complain about this (I guess it's the same behavior when you try to read four values from a three-value tuple?).

I think discarding values is OK, too, as long as we mention this in the documentation somewhere.

**Comment by theodelrieu** [comment]
Yes you get a compile-time error if you go out of bounds with `tuple`.

**Comment by nlohmann** [comment]
In #553, we excluded a conversion from `json` to `std::array`. If we have compile-time checks for tuples, maybe adding support for arrays in the same fashion is also possible?

File: src/json.hpp [hunk 1]
```diff
@@ -1094,6 +1145,25 @@ void from_json(const BasicJsonType& j, ArithmeticType& val)
     }
 }
 
+template <typename BasicJsonType, typename... Args>
+void from_json(const BasicJsonType& j, std::pair<Args...>& p)
+{
+    // should we check that j.size() exactly matches the number of arguments for tuple and pairs?
```
Comment (nlohmann) [inline src/json.hpp]: No, it's OK not to check this. The comment can be removed.

File: test/src/unit-constructor1.cpp [hunk 2]
```diff
@@ -239,6 +239,31 @@ TEST_CASE("constructors")
...
+
+            CHECK(j.type() == json::value_t::array);
+            REQUIRE(j.size() == 4);
+            CHECK(j[0] == std::get<0>(t));
+            CHECK(j[1] == std::get<1>(t));
+            CHECK(j[2] == std::get<2>(t));
+            CHECK(j[3][0] == 0);
+            CHECK(j[3][1] == 1);
+        }
+
```
Comment (nlohmann) [inline test/src/unit-constructor1.cpp]: Could you also add a test case for the case when we have an array with, say, 3 elements and ask for a `std::pair` and something similar with tuples?

Comment (theodelrieu) [inline test/src/unit-constructor1.cpp]: Here it is, sorry for the delay.

**Comment by theodelrieu** [comment]
I will take a look at `std::array` right now.

**Comment by nlohmann** [comment]
@theodelrieu Yes, a nice exception would definitely help. I guess most people would never expect `at` to be called in the first place.

And yes, the pre/post conditions should be documented. I think the README paragraph is the best place to start.

**Comment by nlohmann** [comment]
Thanks a lot, @theodelrieu !

**Pull request merged**